app/Models/Beatmapset.php

```diff
@@ -661,6 +686,82 @@ public function beatmaps()
         return $this->hasMany(Beatmap::class, 'beatmapset_id');
     }
 
+    public function events()
+    {
+        return $this->hasMany(BeatmapsetEvent::class);
+    }
+
+    public function requiredNominationCount()
+    {
+        $longest_map_duration = $this->beatmaps->max('total_length');
+
+        return $longest_map_duration > 315 ? 3 : 2;
+    }
+
+    public function currentNominationCount()
+    {
+        return count($this->recentEvents()->nominations()->get());
+    }
+
+    public function rankingETA()
+    {
+        if (!$this->isQualified()) {
+            return;
+        }
+
+        $queueSize = static::qualified()->where('approved_date', '<', $this->approved_date)->count();
+        $days = ceil($queueSize / static::QUALIFICATIONS_PER_DAY);
+
+        return $days > 0 ? Carbon::now()->addDays($days)->startOfDay() : null;
+    }
+
+    public function recentEvents()
+    {
+        // relevant events differ depending on state of beatmapset
+        $events = null;
+        switch ($this->approved) {
+            case self::PENDING:
+            case self::QUALIFIED:
+                // last 'disqualify' event (if any) and all events since
+                $disqualifyEvent = $this->events()->disqualifications()->orderBy('created_at', 'desc')->first();
```

Comment:
just don't do `$events->discualifications()->...` here, it'll result in sadness.